guides/inventing-service.md

```diff
@@ -0,0 +1,866 @@
+        Box::pin(async move {
+            // same implementation as we had before
+            if request.path() == "/" {
+                Ok(Response::ok("Hello, World!"))
+            } else if request.path() == "/important-data" {
+                let some_data = fetch_data_from_database().await?;
+                Ok(make_response(some_data))
+            } else {
+                Ok(Response::not_found())
+            }
+        })
+    }
+}
+```
+
+How about supporting timeouts? Remember, the solution we're aiming for is one
+that allows us to combine different pieces of functionality together without
+having to modify each individual piece.
+
+What if we defined a generic `Timeout` struct like this:
+
+```rust
+struct Timeout<T> {
+    // T will be some type that implements `Handler`
+    inner_handler: T,
+    duration: Duration,
+}
+```
+
+We could then implement `Handler` for `Timeout<T>` and delegate to `T`'s
+`Handler` implementation:
+
+```rust
+impl<T> Handler for Timeout<T>
+where
+    T: Handler,
+{
+    type Future = Pin<Box<dyn Future<Output = Result<Response, Error>>>>;
+
+    fn call(&mut self, request: Request) -> Self::Future {
+        Box::pin(async move {
+            let result = tokio::time::timeout(
+                self.duration,
+                self.inner_handler.call(request),
+            ).await;
+
+            match result {
+                Ok(Ok(response)) => Ok(response),
+                Ok(Err(error)) => Err(error),
+                Err(_timeout) => Err(Error::timeout()),
+            }
+        })
+    }
+}
+```
+
+The important line being `self.inner_handler.call(request)`. This is where we
+delegate to the inner handler and let it do its thing. We don't know what it is
+we just know that it produces a `Result<Response, Error>` when its done.
+
+This code doesn't quite compile though. We get an error like this:
+
+```
+error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
+   --> src/lib.rs:145:29
+    |
+144 |       fn call(&mut self, request: Request) -> Self::Future {
+    |               --------- this data with an anonymous lifetime `'_`...
+145 |           Box::pin(async move {
+    |  _____________________________^
+146 | |             let result = tokio::time::timeout(
+147 | |                 self.duration,
+148 | |                 self.inner_handler.call(request),
+...   |
+155 | |             }
+156 | |         })
+    | |_________^ ...is captured here, requiring it to live as long as `'static`
+```
+
+The issue is that we're capturing a `&mut self` and moving it into an async
+block. That means the lifetime of our future is tied to the lifetime of `&mut
+self`. This doesn't work for us since we might want to run our response futures
+on multiple threads to get better performance or produce multiple response
+futures and run them all in parallel. That isn't possible if a reference to the
```

Comment:
This isn't actually the reason for this error, basically, TLDR we don't have GAT. All associated types must return a `'static` reference and since `'static` > `'a`  from `&mut 'a self` it doesn't like it. Async-trait gets around this above as I mentioned in that comment, plus the fact that it is a trait object so a concrete type vs a generic like we have here.

Reply:
Yes thats totally correct. Again I'm not sure how much lifetime trouble to actually go into. I'm a bit worried that if we say `Pin<Box<dyn Future>>` needs a `+ 'a` from `&'a mut self` then some users might wonder "well easy, just define service as `trait Service<'a>`". For example this does compile:

```rust
use std::future::Future;

trait Service<'a, Request> {
    type Response;
    type Error;
    type Future: Future<Output = Result<Self::Response, Self::Error>> + 'a;

    fn call(&'a mut self, request: Request) -> Self::Future;
}

struct HttpRequest;
struct Error;

async fn call_all<S>(mut svc: S, requests: Vec<HttpRequest>) -> Result<(), Error>
where
    for<'a> S: Service<'a, HttpRequest, Error = Error>,
{
    for request in requests {
        svc.call(request).await?;
    }
    
    Ok(())
}
```

But we can't `tokio::spawn` the response as that requires the future to be `'static` which then makes `self` `&'static mut self` which clearly is quite limiting.

So yeah you're totally correct but I'm trying to say something that will make users think "yeah that kinda makes sense" without opening a can of worms 😅

What do you think?

Reply:
I think you're right there is a balance to be achieved here 😄

Reply:
I think it's worth leaving a note here saying that `clone` isn't the only way to do this, and that in the future we're hoping to be able to express this in an even cleaner way (link to GAT issue). That's probably enough for the casual reader.

Reply:
Adding it as a footnote is a good idea!